# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/xul/nsBoxFrame.cpp
# Commit: 4b5f53d5127b
# Full Hash: 4b5f53d5127b36ac51eb71c678b1aa126a066c79
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2021-05-18 09:45:31
# Regressor Bug: 1540737
# File Overlap Count: 1
# Description:
#   Bug 1540737 - Implement Paint() for all display items, and nsDisplayList so that we can paint without using FrameLayerBuilder. r=miko
#   
#   This creates a new leaf nsDisplayWrapper item for consumers that just want an empty wrapper, with a Paint method that asserts. This lets us leaves Paint pure-virtual on nsDisplayWrapList, and lets the compiler force us to implement it for the other leaf types.
#   
#   It also disables flattening of transform/opacity items. This is because they flatten the items even when used with FlattenedDisplayItemIterator, even though this doesn't add markers to replace them (only FLB does this). We could in theory fix this, but I don't think FLB performance matters much any more.
# ==============================================================================

diff -r a73e041f1f1d -r 4b5f53d5127b layout/xul/nsBoxFrame.cpp
--- a/layout/xul/nsBoxFrame.cpp	Mon May 17 23:53:44 2021 +0000
+++ b/layout/xul/nsBoxFrame.cpp	Mon May 17 23:53:45 2021 +0000
@@ -1043,6 +1043,10 @@
   virtual bool ShouldFlattenAway(nsDisplayListBuilder* aBuilder) override {
     return false;
   }
+  void Paint(nsDisplayListBuilder* aBuilder, gfxContext* aCtx) override {
+    GetChildren()->Paint(aBuilder, aCtx,
+                         mFrame->PresContext()->AppUnitsPerDevPixel());
+  }
   NS_DISPLAY_DECL_NAME("XULEventRedirector", TYPE_XUL_EVENT_REDIRECTOR)
  private:
   nsIFrame* mTargetFrame;
@@ -1083,7 +1087,7 @@
   }
 }
 
-class nsXULEventRedirectorWrapper final : public nsDisplayWrapper {
+class nsXULEventRedirectorWrapper final : public nsDisplayItemWrapper {
  public:
   explicit nsXULEventRedirectorWrapper(nsIFrame* aTargetFrame)
       : mTargetFrame(aTargetFrame) {}
